### PR TITLE
Option to display notifications listing CSP violations

### PR DESCRIPTION
**Pull request by daniil-berg**
## Purpose
We want to be able to notify (or even consistently nag) our teachers when they embed external content from questionable sources in their courses, rather than merely blocking it.

## Main functionality
This branch adds an option to the plugin settings to always display a [notification](https://docs.moodle.org/dev/Notifications) to the user, whenever security policy violations occur on the page he is visiting, in which the sources (URIs) in question are listed.
Example:
example_notification

## Implementation details
This is accomplished by listening to the [SecurityPolicyViolation](https://developer.mozilla.org/en-US/docs/Web/API/SecurityPolicyViolationEvent) Events issued when external content is blocked (or a violation is only reported).

To allow role-based control over what users will be shown such a notification, a new capability is introduced to the course context. (By default, it is not shown to students, but is shown to teachers.)

There is also an option (dependent on enabling the notifications at all), which determines if only blocked sources should be shown in the notification. This means, for instance, if that option is on and a page only has violations that are reported (but not blocked), the notification does not come up at all.

The entire JavaScript hooks into Moodle's [`before_standard_html_head`](https://docs.moodle.org/dev/Output_callbacks#before_standard_html_head) callback, to make sure that the event listener is in place before the events are triggered.

A simple template file is used for the notification body. The text therein as well as the new settings and the capability all make use of translatable strings.

## Related
This is intended to supersede #64 as a more elegant solution.

### CODE REVIEW

File: lang/en/local_csp.php [hunk 1]
```diff
@@ -69,3 +69,14 @@
...
 $string['violateddirective'] = 'Policy';
 $string['privacy:metadata'] = 'The CSP plugin contains no user specific data.';
+
+# Notifications
+$string['notificationheader'] = 'Potentially insecure content detected!';
+$string['notificationstart'] = 'This page contains embedded content that the administrators consider potentially insecure.';
+$string['notificationblocked'] = 'Content from the following sources has been automatically blocked:';
+$string['notificationnotblocked'] = 'Content from the following sources was found (but has not been blocked):';
+$string['notificationsenable'] = 'Display violation notifications';
+$string['notificationsenabledescription'] = 'Display notifications to the user, when CSP violations are detected on the page he visits. They will list all the blocked URIs on that page.';
```
Comment (brendanheywood) [inline lang/en/local_csp.php]: Can you please make this gender neutral eg

Display notifications to the user, when CSP violations are detected on the current page.

**Comment by brendanheywood** [comment]
hi @daniil-berg this looks really cool. A few high level thing things:

1) When I tested this I could get the warning notification but I didn't get it to show the list of urls which were blocked. This was with both notifications_enforced_only on and off and tested in chrome and firefox.

2) I found the two admin checkboxes a little confusing, partially because of the conditional hiding, and perhaps also because the description might need some tweaking. But I was wondering whether it would be better to make them properly independent and always have both checkboxes available. ie I mean if you have a real policy in place that blocks A, and you also have a reporting policy which blocks B, then you would get an **error** notification about A and a **warning** notification about B. If you do go down that path I'd try to make sure the errors appear at the top above the warning. Then you would have each checkbox setting grouped with it's policy textarea.

**Comment by brendanheywood** [comment]
Also I'm wondering if it is worth making the hard coded 1000ms an admin setting. I don't know what is the ideal here but I can imagine some sites might want to tune it.

**Comment by daniil-berg** [comment]
Hey Brendan, thanks for checking this out. 

I'll address the most pressing matter first: The fact that you can't see the URI list. It was working for me with no issues so far, both on Chromium and Firefox.

I am attaching a small [patch](https://github.com/catalyst/moodle-local_csp/files/8443379/patch.tar.gz) that only adds some verbosity to the console log. I hope this will help us figure out, what is going on. Please apply it, make sure to purge all the Moodle caches at `admin/purgecaches.php`, and try again. The log should contain an entry for each violation event caught (relatively early on), then a message about how many events were appended to the `localCspViolations` array and passed to the `displayNotification` function, and finally the number of enforced/reported violations.

Please let me know, at which point something unexpected occurs.

**EDIT:** Just found out that apparently some browser extensions such as uBlock Origin are able to block CSP reports. It may make sense to check that none of your extensions interfere. (Also, just tested it again and it works on mobile for me as well.)

**Comment by brendanheywood** [comment]
Can you please push that as another commit or another branch?

**Comment by brendanheywood** [comment]
Nevermind, got this working, playing now to get some logs for you

**Comment by brendanheywood** [comment]
The csp reports are coming fine, I can see all the details, and your custom logging is working:

image

debugging I can see all the raw data:
image

image

I think one part of the problem is enforcedOnly === false, not enforcedOnly === 0

**Comment by daniil-berg** [comment]
What about that last log message about the number of enforced/reported events? I don't see it in your screenshots.

EDIT: Also, even if your suggestion holds, this should not affect the construction of the `enforcedEvents` array. Can you try with some enforced events too?

EDIT 2: You are absolutely right though. That line should be:
```javascript
} else if (enforcedOnly === false && event.disposition === 'report') {
```
This breaks displaying the reported ones. I will fix that of course. But I hope you get the enforced ones showing, at least?

**Comment by brendanheywood** [comment]
Ok just tried a simpler test with these settings:

image

and I get this:
image

but with these settings:
image

I get this:
image

It's 1am so I need to get to sleep but I hope thats enough to reproduce and track it down.

**Comment by daniil-berg** [comment]
Sorry, forgot about the timezone difference. :wink: 
I implemented most of what you've mentioned here. I hope it works on your end now.

The order of the warning/error notices is currently not guaranteed. This is because I see no way to definitively decide, when to stop waiting for a potential error notice, before displaying a warning or vice-versa. At least without introducing another arbitrary timeout/delay. I personally don't think it makes much of a difference in what order these notifications are displayed.

The other timeout that is attached to the event listener waiting to display the collected violations is still hard-coded at 1 second (but is now set in a different place). I attached a TODO comment for us to think this over. I also thought about making this configurable, but this seems like overkill at this point.

I am glad you like the proposed additions.
Unless you find any more bugs or insist on crucial functionality additions/changes, I would appreciate if you could merge this. I think we can still build on it afterwards (potentially adding the timeout as an admin setting or other non-essential changes).

Cheers,
Dan

**Comment by brendanheywood** [comment]
thanks @daniil-berg this is a great little addition. The only extra thing I've done is default the error notification to off

**Pull request merged**